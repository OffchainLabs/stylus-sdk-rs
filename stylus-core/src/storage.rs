// Copyright 2024-2025, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

/// Trait for top-level storage types, usually implemented by proc macros.
/// Top-level types are special in that their lifetimes track the entirety
/// of all the EVM state-changes throughout a contract invocation.
///
/// To prevent storage aliasing during reentrancy, you must hold a reference
/// to such a type when making an EVM call. This may change in the future
/// for programs that prevent reentrancy.
///
/// # Safety
///
/// The type must be top-level to prevent storage aliasing.
pub unsafe trait TopLevelStorage {}
