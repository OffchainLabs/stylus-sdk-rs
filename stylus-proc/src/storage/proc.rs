// Copyright 2023-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use lazy_static::lazy_static;
use proc_macro2::{Ident, Literal};
use quote::quote;
use regex::Regex;
use syn::{
    braced, bracketed, parenthesized,
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    token::Bracket,
    Attribute, Error, Generics, Path, Result, Token, Visibility,
};

macro_rules! sdk {
    ($($msg:expr),+) => {
        format!("stylus_sdk::storage::{}", format!($($msg),+))
    };
}

pub struct SolidityStructs(pub Vec<SolidityStruct>);

impl Parse for SolidityStructs {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut structs = Vec::new();
        while !input.is_empty() {
            structs.push(input.parse()?);
        }
        Ok(Self(structs))
    }
}

pub struct SolidityStruct {
    pub attrs: Vec<Attribute>,
    pub vis: Visibility,
    pub name: Ident,
    pub generics: Generics,
    pub fields: SolidityFields,
}

impl Parse for SolidityStruct {
    fn parse(input: ParseStream) -> Result<Self> {
        // #[attrs?]
        // pub? struct name
        let attrs: Vec<Attribute> = Attribute::parse_outer(input)?;
        let vis: Visibility = input.parse()?;
        let _: Token![struct] = input.parse()?;
        let name: Ident = input.parse()?;
        let generics: Generics = input.parse()?;

        let content;
        let _ = braced!(content in input);
        let fields = content.parse()?;
        Ok(Self {
            attrs,
            vis,
            name,
            generics,
            fields,
        })
    }
}

pub struct SolidityFields(pub Punctuated<SolidityField, Token![;]>);

impl Parse for SolidityFields {
    fn parse(input: ParseStream) -> Result<Self> {
        let fields = Punctuated::parse_terminated(input)?;
        Ok(Self(fields))
    }
}

pub struct SolidityField {
    pub attrs: Vec<Attribute>,
    pub name: Ident,
    pub ty: Path,
}

impl Parse for SolidityField {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs: Vec<Attribute> = Attribute::parse_outer(input)?;
        let ty = SolidityTy::parse(input)?.0;
        let name: Ident = input.parse()?;
        Ok(SolidityField { attrs, name, ty })
    }
}

pub struct SolidityTy(Path);

impl Parse for SolidityTy {
    fn parse(input: ParseStream) -> Result<Self> {
        let start: Path = input.parse()?;
        let mut path: Path;

        if start.is_ident("mapping") {
            let content;
            let _ = parenthesized!(content in input);

            // key => value
            let key = content.parse::<PrimitiveKey>()?.0;
            let _: Token![=>] = content.parse()?;
            let value = content.parse::<SolidityTy>()?.0;

            let ty = format!(
                "{}<{}, {}>",
                sdk!("StorageMap"),
                quote!(#key),
                quote!(#value)
            );
            path = syn::parse_str(&ty)?;
        } else {
            let base: Primitive = start.try_into()?;
            path = base.0
        };

        while input.peek(Bracket) {
            let content;
            let _ = bracketed!(content in input);

            if content.is_empty() {
                let outer = sdk!("StorageVec");
                let inner = quote! { #path };
                path = syn::parse_str(&format!("{outer}<{inner}>"))?;
            } else {
                let content: Literal = content.parse()?;
                let Ok(size) = content.to_string().parse::<usize>() else {
                    error!(@content, "Array size must be a positive integer");
                };
                let outer = sdk!("StorageArray");
                let inner = quote! { #path };
                path = syn::parse_str(&format!("{outer}<{inner}, {size}>"))?;
            }
        }

        Ok(SolidityTy(path))
    }
}

pub struct Primitive(Path);

lazy_static! {
    static ref UINT_REGEX: Regex = Regex::new(r"^uint(\d+)$").unwrap();
    static ref INT_REGEX: Regex = Regex::new(r"^int(\d+)$").unwrap();
    static ref BYTES_REGEX: Regex = Regex::new(r"^bytes(\d+)$").unwrap();
    static ref LOWER_REGEX: Regex = Regex::new(r"^[0-9a-z]+$").unwrap();
}

impl Parse for Primitive {
    fn parse(input: ParseStream) -> Result<Self> {
        let path: Path = input.parse()?;
        path.try_into()
    }
}

impl TryFrom<Path> for Primitive {
    type Error = Error;

    fn try_from(path: Path) -> std::result::Result<Self, Self::Error> {
        let Some(ident) = path.get_ident() else {
            return Ok(Self(path));
        };
        let name = &ident.to_string();

        macro_rules! ty {
            ($($msg:expr),+) => {{
                let path = sdk!($($msg),+);
                Ok(Self(syn::parse_str(&path)?))
            }};
        }
        macro_rules! error {
            ($msg:expr) => {
                Err(Error::new_spanned(&ident, $msg))
            };
        }

        if let Some(caps) = UINT_REGEX.captures(name) {
            let bits: usize = caps[1].parse().unwrap();
            let limbs = (63 + bits) / 64;
            if bits > 256 {
                return error!("Type not supported: too many bits");
            }
            return ty!("StorageUint<{}, {}>", bits, limbs);
        }

        if let Some(caps) = INT_REGEX.captures(name) {
            let bits: usize = caps[1].parse().unwrap();
            let limbs = (63 + bits) / 64;
            if bits > 256 {
                return error!("Type not supported: too many bits");
            }
            return ty!("StorageSigned<{}, {}>", bits, limbs);
        }

        if let Some(caps) = BYTES_REGEX.captures(name) {
            let bytes: usize = caps[1].parse().unwrap();
            if bytes > 32 {
                return error!("Type not supported: too many bytes");
            }
            return ty!("StorageFixedBytes<{}>", bytes);
        }

        let ty = match name.as_str() {
            "address" => "StorageAddress",
            "bool" => "StorageBool",
            "bytes" => "StorageBytes",
            "int" => "StorageI256",
            "string" => "StorageString",
            "uint" => "StorageU256",
            x => match LOWER_REGEX.is_match(x) {
                true => return Err(Error::new_spanned(ident, "Type not supported")),
                false => return Ok(Self(syn::parse_str(x)?)),
            },
        };

        ty!("{ty}")
    }
}

pub struct PrimitiveKey(Path);

impl Parse for PrimitiveKey {
    fn parse(input: ParseStream) -> Result<Self> {
        let path: Path = input.parse()?;
        path.try_into()
    }
}

impl TryFrom<Path> for PrimitiveKey {
    type Error = Error;

    fn try_from(path: Path) -> std::result::Result<Self, Self::Error> {
        let Some(ident) = path.get_ident() else {
            return Ok(Self(path));
        };
        let name = &ident.to_string();

        macro_rules! ty {
            ($($msg:expr),+) => {{
                let path = format!($($msg),+);
                let path = format!("stylus_sdk::alloy_primitives::{path}");
                Ok(Self(syn::parse_str(&path)?))
            }};
        }
        macro_rules! error {
            ($msg:expr) => {
                Err(Error::new_spanned(&ident, $msg))
            };
        }

        if let Some(caps) = UINT_REGEX.captures(name) {
            let bits: usize = caps[1].parse().unwrap();
            let limbs = (63 + bits) / 64;
            if bits > 256 {
                return error!("Type not supported: too many bits");
            }
            return ty!("Uint<{}, {}>", bits, limbs);
        }

        if let Some(caps) = INT_REGEX.captures(name) {
            let bits: usize = caps[1].parse().unwrap();
            let limbs = (63 + bits) / 64;
            if bits > 256 {
                return error!("Type not supported: too many bits");
            }
            return ty!("Signed<{}, {}>", bits, limbs);
        }

        if let Some(caps) = BYTES_REGEX.captures(name) {
            let bytes: usize = caps[1].parse().unwrap();
            if bytes > 32 {
                return error!("Type not supported: too many bytes");
            }
            return ty!("FixedBytes<{}>", bytes);
        }

        let ty = match name.as_str() {
            "address" => "Address",
            "bool" => "U8",
            "int" => "I256",
            "uint" => "U256",
            "bytes" => return Ok(Self(syn::parse_str("Vec<u8>")?)),
            "string" => return Ok(Self(syn::parse_str("String")?)),
            x => match LOWER_REGEX.is_match(x) {
                true => return Err(Error::new_spanned(ident, "Type not supported")),
                false => return Ok(Self(syn::parse_str(x)?)),
            },
        };

        ty!("{ty}")
    }
}
