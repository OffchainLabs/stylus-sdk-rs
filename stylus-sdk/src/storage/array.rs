// Copyright 2023-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use super::{Erase, StorageGuard, StorageGuardMut, StorageType};
use alloy_primitives::U256;
use core::marker::PhantomData;

/// Accessor for a storage-backed array.
pub struct StorageArray<S: StorageType, const N: usize> {
    slot: U256,
    marker: PhantomData<S>,
}

impl<S: StorageType, const N: usize> StorageType for StorageArray<S, N> {
    type Wraps<'a> = StorageGuard<'a, StorageArray<S, N>> where Self: 'a;
    type WrapsMut<'a> = StorageGuardMut<'a, StorageArray<S, N>> where Self: 'a;

    const REQUIRED_SLOTS: usize = Self::required_slots();

    unsafe fn new(slot: U256, offset: u8) -> Self {
        debug_assert!(offset == 0);
        Self {
            slot,
            marker: PhantomData,
        }
    }

    fn load<'s>(self) -> Self::Wraps<'s> {
        StorageGuard::new(self)
    }

    fn load_mut<'s>(self) -> Self::WrapsMut<'s> {
        StorageGuardMut::new(self)
    }
}

impl<S: StorageType, const N: usize> StorageArray<S, N> {
    /// Gets the number of elements stored.
    ///
    /// Although this type will always have the same length, this method is still provided for
    /// consistency with [`StorageVec`].
    pub const fn len(&self) -> usize {
        N
    }

    /// Gets an accessor to the element at a given index, if it exists.
    /// Note: the accessor is protected by a [`StorageGuard`], which restricts
    /// its lifetime to that of `&self`.
    pub fn getter(&self, index: impl TryInto<usize>) -> Option<StorageGuard<S>> {
        let store = unsafe { self.accessor(index)? };
        Some(StorageGuard::new(store))
    }

    /// Gets a mutable accessor to the element at a given index, if it exists.
    /// Note: the accessor is protected by a [`StorageGuardMut`], which restricts
    /// its lifetime to that of `&mut self`.
    pub fn setter(&mut self, index: impl TryInto<usize>) -> Option<StorageGuardMut<S>> {
        let store = unsafe { self.accessor(index)? };
        Some(StorageGuardMut::new(store))
    }

    /// Gets the underlying accessor to the element at a given index, if it exists.
    ///
    /// # Safety
    ///
    /// Enables aliasing.
    unsafe fn accessor(&self, index: impl TryInto<usize>) -> Option<S> {
        let index = index.try_into().ok()?;
        if index >= N {
            return None;
        }
        let (slot, offset) = self.index_slot(index);
        Some(S::new(slot, offset))
    }

    /// Gets the underlying accessor to the element at a given index, even if out of bounds.
    ///
    /// # Safety
    ///
    /// Enables aliasing. UB if out of bounds.
    unsafe fn accessor_unchecked(&self, index: usize) -> S {
        let (slot, offset) = self.index_slot(index);
        S::new(slot, offset)
    }

    /// Gets the element at the given index, if it exists.
    pub fn get(&self, index: impl TryInto<usize>) -> Option<S::Wraps<'_>> {
        let store = unsafe { self.accessor(index)? };
        Some(store.load())
    }

    /// Gets a mutable accessor to the element at a given index, if it exists.
    pub fn get_mut(&mut self, index: impl TryInto<usize>) -> Option<S::WrapsMut<'_>> {
        let store = unsafe { self.accessor(index)? };
        Some(store.load_mut())
    }

    /// Determines the slot and offset for the element at an index.
    fn index_slot(&self, index: usize) -> (U256, u8) {
        let width = S::SLOT_BYTES;
        let words = S::REQUIRED_SLOTS.max(1);
        let density = Self::density();

        let slot = self.slot + U256::from(words * index / density);
        let offset = 32 - (width * (1 + index % density)) as u8;
        (slot, offset)
    }

    /// Number of elements per slot.
    const fn density() -> usize {
        32 / S::SLOT_BYTES
    }

    /// Required slots for the storage array.
    const fn required_slots() -> usize {
        let reserved = N * S::REQUIRED_SLOTS;
        let density = Self::density();
        let packed = (N + density - 1) / density; // ceil division for packed items.
        if reserved > packed {
            return reserved;
        }
        packed
    }
}

impl<S: Erase, const N: usize> Erase for StorageArray<S, N> {
    fn erase(&mut self) {
        for i in 0..N {
            let mut store = unsafe { self.accessor_unchecked(i) };
            store.erase()
        }
    }
}
