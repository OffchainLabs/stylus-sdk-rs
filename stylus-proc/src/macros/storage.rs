// Copyright 2023-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use proc_macro2::TokenStream;
use proc_macro_error::emit_error;
use quote::{quote, ToTokens};
use syn::{
    parse_macro_input, parse_quote, punctuated::Punctuated, spanned::Spanned, ItemStruct, Token,
    Type,
};

use crate::{consts::STYLUS_HOST_FIELD, utils::attrs::consume_flag};

/// Implementation of the [`#[storage]`][crate::storage] macro.
pub fn storage(
    attr: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    if !attr.is_empty() {
        emit_error!(
            TokenStream::from(attr).span(),
            "this macro is not configurable"
        );
    }

    let item = parse_macro_input!(input as ItemStruct);
    let ItemStruct {
        attrs,
        vis,
        ident,
        generics,
        fields,
        ..
    } = item;

    // Handle fields based on their type (named or unnamed)
    let expanded_fields = match fields {
        syn::Fields::Named(named_fields) => {
            // Extract the original fields.
            let original_fields = named_fields.named;
            quote! {
                #STYLUS_HOST_FIELD: stylus_sdk::host::VM,
                #original_fields
            }
        }
        syn::Fields::Unnamed(_) => {
            // Handle tuple structs if needed.
            emit_error!(
                fields.span(),
                "Tuple structs are not supported by #[storage]"
            );
            return fields.to_token_stream().into();
        }
        syn::Fields::Unit => {
            // Handle unit structs if needed.
            quote! {
                #STYLUS_HOST_FIELD: stylus_sdk::host::VM,
            }
        }
    };
    // Inject the host trait generic into the item struct if not defined.
    let mut host_injected_item: syn::ItemStruct = parse_quote! {
        #(#attrs)*
        #vis struct #ident #generics {
            #expanded_fields
        }
    };
    let storage = Storage::from(&mut host_injected_item);
    let mut output = host_injected_item.into_token_stream();
    storage.to_tokens(&mut output);
    output.into()
}

#[derive(Debug)]
struct Storage {
    name: syn::Ident,
    generics: syn::Generics,
    fields: Vec<StorageField>,
}

impl Storage {
    fn item_impl(&self) -> syn::ItemImpl {
        let name = &self.name;
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        let size = TokenStream::from_iter(self.fields.iter().map(StorageField::size));
        parse_quote! {
            impl #impl_generics #name #ty_generics #where_clause {
                const fn required_slots() -> usize {
                    use stylus_sdk::storage;
                    let mut total: usize = 0;
                    let mut space: usize = 32;
                    #size
                    if space != 32 || total == 0 {
                        total += 1;
                    }
                    total
                }
            }
        }
    }

    fn impl_storage_type(&self) -> syn::ItemImpl {
        let name = &self.name;
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        let init = Punctuated::<syn::FieldValue, Token![,]>::from_iter(
            self.fields.iter().filter_map(StorageField::init),
        );
        parse_quote! {
            impl #impl_generics stylus_sdk::storage::StorageType for #name #ty_generics #where_clause {
                type Wraps<'a> = stylus_sdk::storage::StorageGuard<'a, Self> where Self: 'a;
                type WrapsMut<'a> = stylus_sdk::storage::StorageGuardMut<'a, Self> where Self: 'a;

                // start a new word
                const SLOT_BYTES: usize = 32;
                const REQUIRED_SLOTS: usize = Self::required_slots();

                unsafe fn new(mut root: stylus_sdk::alloy_primitives::U256, offset: u8, host: stylus_sdk::host::VM) -> Self {
                    use stylus_sdk::{storage, alloy_primitives};
                    debug_assert!(offset == 0);

                    let mut space: usize = 32;
                    let mut slot: usize = 0;
                    let accessor = Self {
                        #STYLUS_HOST_FIELD: host.clone(),
                        #init
                    };
                    accessor
                }

                fn load<'s>(self) -> Self::Wraps<'s> {
                    stylus_sdk::storage::StorageGuard::new(self)
                }

                fn load_mut<'s>(self) -> Self::WrapsMut<'s> {
                    stylus_sdk::storage::StorageGuardMut::new(self)
                }
            }
        }
    }

    fn impl_host_access(&self) -> syn::ItemImpl {
        let name = &self.name;
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        parse_quote! {
            impl #impl_generics stylus_sdk::stylus_core::HostAccess for #name #ty_generics #where_clause {
                fn vm(&self) -> &dyn stylus_sdk::stylus_core::Host {
                    #[cfg(target_arch = "wasm32")]
                    {
                        &self.__stylus_host
                    }
                    #[cfg(not(target_arch = "wasm32"))]
                    {
                        self.__stylus_host.host.as_ref()
                    }
                }
            }
        }
    }
    fn impl_from_vm(&self) -> syn::ItemImpl {
        let name = &self.name;
        // Use a more unique name for the Host generic
        let (_, ty_generics, where_clause) = self.generics.split_for_impl();
        let mut new_generics = self.generics.clone();
        let host_param = parse_quote!(__StylusHostType: stylus_sdk::stylus_core::Host + 'static);
        new_generics.params.push(host_param);
        let (impl_generics, _, _) = new_generics.split_for_impl();

        parse_quote! {
            #[cfg(not(target_arch = "wasm32"))]
            impl #impl_generics From<__StylusHostType> for #name #ty_generics #where_clause {
                fn from(host: __StylusHostType) -> Self {
                    unsafe {
                        Self::new(
                            stylus_sdk::alloy_primitives::U256::ZERO,
                            0,
                            stylus_sdk::host::VM {
                                host: alloc::boxed::Box::new(host),
                            },
                        )
                    }
                }
            }
        }
    }
}

impl From<&mut syn::ItemStruct> for Storage {
    fn from(node: &mut syn::ItemStruct) -> Self {
        let name = node.ident.clone();
        let generics = node.generics.clone();
        let fields = node
            .fields
            .iter_mut()
            .enumerate()
            .filter_map(|(idx, field)| {
                if let syn::Type::Path(..) = &field.ty {
                    Some(StorageField::new(idx, field))
                } else {
                    None
                }
            })
            .collect();
        Self {
            name,
            generics,
            fields,
        }
    }
}

impl ToTokens for Storage {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        self.item_impl().to_tokens(tokens);
        self.impl_storage_type().to_tokens(tokens);
        self.impl_host_access().to_tokens(tokens);
        self.impl_from_vm().to_tokens(tokens);
        for field in &self.fields {
            field.impl_borrow(&self.name).to_tokens(tokens);
            field.impl_borrow_mut(&self.name).to_tokens(tokens);
        }
    }
}

#[derive(Debug)]
struct StorageField {
    name: Option<syn::Ident>,
    ty: syn::Type,
    accessor: syn::Member,
    borrow: bool,
}

impl StorageField {
    fn new(idx: usize, field: &mut syn::Field) -> Self {
        check_type(field);
        let name = field.ident.clone();
        let ty = field.ty.clone();
        let accessor = field
            .ident
            .clone()
            .map(syn::Member::from)
            .unwrap_or_else(|| idx.into());

        let borrow = consume_flag(&mut field.attrs, "borrow");
        Self {
            name,
            ty,
            accessor,
            borrow,
        }
    }

    fn init(&self) -> Option<syn::FieldValue> {
        let Some(ident) = &self.name else {
            return None;
        };
        if *ident == STYLUS_HOST_FIELD.as_ident() {
            return None;
        }
        let ty = &self.ty;
        Some(parse_quote! {
            #ident: {
                let bytes = <#ty as storage::StorageType>::SLOT_BYTES;
                let words = <#ty as storage::StorageType>::REQUIRED_SLOTS;
                if space < bytes {
                    space = 32;
                    slot += 1;
                }
                space -= bytes;

                let root = root + alloy_primitives::U256::from(slot);
                let field = <#ty as storage::StorageType>::new(root, space as u8, host.clone());
                if words > 0 {
                    slot += words;
                    space = 32;
                }
                field
            }
        })
    }

    fn size(&self) -> TokenStream {
        let ty = &self.ty;
        let Some(ident) = &self.name else {
            return quote! {};
        };
        if *ident == STYLUS_HOST_FIELD.as_ident() {
            return quote! {};
        }
        quote! {
            let bytes = <#ty as storage::StorageType>::SLOT_BYTES;
            let words = <#ty as storage::StorageType>::REQUIRED_SLOTS;

            if words > 0 {
                total += words;
                space = 32;
            } else {
                if space < bytes {
                    space = 32;
                    total += 1;
                }
                space -= bytes;
            }
        }
    }

    fn impl_borrow(&self, name: &syn::Ident) -> Option<syn::ItemImpl> {
        let Self { ty, accessor, .. } = self;
        self.borrow.then(|| {
            parse_quote! {
                impl core::borrow::Borrow<#ty> for #name {
                    fn borrow(&self) -> &#ty {
                        &self.#accessor
                    }
                }
            }
        })
    }

    fn impl_borrow_mut(&self, name: &syn::Ident) -> Option<syn::ItemImpl> {
        let Self { ty, accessor, .. } = self;
        self.borrow.then(|| {
            parse_quote! {
                impl core::borrow::BorrowMut<#ty> for #name {
                    fn borrow_mut(&mut self) -> &mut #ty {
                        &mut self.#accessor
                    }
                }
            }
        })
    }
}

fn check_type(field: &syn::Field) {
    let Type::Path(ty) = &field.ty else {
        unreachable!();
    };

    let path = &ty.path.segments.last().unwrap().ident;
    let not_supported = format!("Type `{path}` not supported for EVM state storage");

    match path.to_string().as_str() {
        x @ ("u8" | "u16" | "u32" | "u64" | "u128" | "i8" | "i16" | "i32" | "i64" | "i128"
        | "U8" | "U16" | "U32" | "U64" | "U128" | "I8" | "I16" | "I32" | "I64" | "I128") => {
            emit_error!(
                &field,
                "{not_supported}. Instead try `Storage{}`.",
                x.to_uppercase()
            );
        }
        "usize" => emit_error!(&field, "{not_supported}."),
        "isize" => emit_error!(&field, "{not_supported}."),
        "bool" => emit_error!(&field, "{not_supported}. Instead try `StorageBool`."),
        "f32" | "f64" => emit_error!(&field, "{not_supported}. Consider fixed-point arithmetic."),
        _ => {}
    }
}
