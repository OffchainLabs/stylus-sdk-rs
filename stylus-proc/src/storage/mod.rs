// Copyright 2023-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use crate::storage::proc::{SolidityField, SolidityFields, SolidityStruct, SolidityStructs};
use proc_macro::TokenStream;
use quote::{quote, ToTokens};
use std::mem;
use syn::{parse_macro_input, punctuated::Punctuated, Index, ItemStruct, Token, Type};

mod proc;

pub fn storage(_attr: TokenStream, input: TokenStream) -> TokenStream {
    let mut input = parse_macro_input!(input as ItemStruct);

    let name = &input.ident;
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();

    let mut init = quote! {};
    let mut size = quote! {};
    let mut borrows = quote! {};

    for (field_index, field) in input.fields.iter_mut().enumerate() {
        // deny complex types
        let Type::Path(ty) = &field.ty else {
            error!(&field, "Type not supported for EVM state storage");
        };

        // implement borrows
        let attrs = mem::take(&mut field.attrs);
        for attr in attrs {
            if !attr.path.is_ident("borrow") {
                field.attrs.push(attr);
                continue;
            }
            if !attr.tokens.is_empty() {
                error!(attr.tokens, "borrow attribute does not take parameters");
            }
            let ty = &field.ty;
            let accessor = match field.ident.as_ref() {
                Some(accessor) => accessor.into_token_stream(),
                None => Index::from(field_index).into_token_stream(),
            };
            borrows.extend(quote! {
                impl core::borrow::Borrow<#ty> for #name {
                    fn borrow(&self) -> &#ty {
                        &self.#accessor
                    }
                }
                impl core::borrow::BorrowMut<#ty> for #name {
                    fn borrow_mut(&mut self) -> &mut #ty {
                        &mut self.#accessor
                    }
                }
            });
        }

        let path = &ty.path.segments.last().unwrap().ident;
        let not_supported = format!("Type `{path}` not supported for EVM state storage");

        match path.to_string().as_str() {
            x @ ("u8" | "u16" | "u32" | "u64" | "u128" | "i8" | "i16" | "i32" | "i64" | "i128"
            | "U8" | "U16" | "U32" | "U64" | "U128" | "I8" | "I16" | "I32" | "I64"
            | "I128") => {
                error!(
                    &field,
                    "{not_supported}. Instead try `Storage{}`.",
                    x.to_uppercase()
                );
            }
            "usize" => error!(&field, "{not_supported}."),
            "isize" => error!(&field, "{not_supported}."),
            "bool" => error!(&field, "{not_supported}. Instead try `StorageBool`."),
            "f32" | "f64" => error!(&field, "{not_supported}. Consider fixed-point arithmetic."),
            _ => {}
        }

        let Some(ident) = &field.ident else {
            continue;
        };

        init.extend(quote! {
            #ident: {
                let bytes = <#ty as storage::StorageType>::SLOT_BYTES;
                let words = <#ty as storage::StorageType>::REQUIRED_SLOTS;
                if space < bytes {
                    space = 32;
                    slot += 1;
                }
                space -= bytes;

                let root = root + alloy_primitives::U256::from(slot);
                let field = <#ty as storage::StorageType>::new(root, space as u8);
                if words > 0 {
                    slot += words;
                    space = 32;
                }
                field
            },
        });

        size.extend(quote! {
            let bytes = <#ty as storage::StorageType>::SLOT_BYTES;
            let words = <#ty as storage::StorageType>::REQUIRED_SLOTS;

            if words > 0 {
                total += words;
                space = 32;
            } else {
                if space < bytes {
                    space = 32;
                    total += 1;
                }
                space -= bytes;
            }
        });
    }

    let expanded = quote! {
        #input

        impl #impl_generics #name #ty_generics #where_clause {
            const fn required_slots() -> usize {
                use stylus_sdk::storage;
                let mut total: usize = 0;
                let mut space: usize = 32;
                #size
                if space != 32 || total == 0 {
                    total += 1;
                }
                total
            }
        }

        impl #impl_generics stylus_sdk::storage::StorageType for #name #ty_generics #where_clause {
            type Wraps<'a> = stylus_sdk::storage::StorageGuard<'a, Self> where Self: 'a;
            type WrapsMut<'a> = stylus_sdk::storage::StorageGuardMut<'a, Self> where Self: 'a;

            // start a new word
            const SLOT_BYTES: usize = 32;
            const REQUIRED_SLOTS: usize = Self::required_slots();

            unsafe fn new(mut root: stylus_sdk::alloy_primitives::U256, offset: u8) -> Self {
                use stylus_sdk::{storage, alloy_primitives};
                debug_assert!(offset == 0);

                let mut space: usize = 32;
                let mut slot: usize = 0;
                let accessor = Self {
                    #init
                };
                accessor
            }

            fn load<'s>(self) -> Self::Wraps<'s> {
                stylus_sdk::storage::StorageGuard::new(self)
            }

            fn load_mut<'s>(self) -> Self::WrapsMut<'s> {
                stylus_sdk::storage::StorageGuardMut::new(self)
            }
        }

        #borrows
    };
    expanded.into()
}

pub fn sol_storage(input: TokenStream) -> TokenStream {
    let SolidityStructs(decls) = parse_macro_input!(input as SolidityStructs);
    let mut out = quote!();

    for decl in decls {
        let SolidityStruct {
            attrs,
            vis,
            name,
            generics,
            fields: SolidityFields(fields),
        } = decl;

        let fields: Punctuated<_, Token![,]> = fields
            .into_iter()
            .map(|SolidityField { attrs, name, ty }| {
                quote! {
                    #(#attrs)*
                    pub #name: #ty
                }
            })
            .collect();

        out.extend(quote! {
            #(#attrs)*
            #[stylus_sdk::stylus_proc::storage]
            #vis struct #name #generics {
                #fields
            }
        });
    }

    out.into()
}

pub fn derive_erase(input: TokenStream) -> TokenStream {
    let mut input = parse_macro_input!(input as ItemStruct);
    let name = &input.ident;
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();

    let mut erase_fields = quote! {};
    for field in &mut input.fields {
        let ident = &field.ident;
        erase_fields.extend(quote! {
            self.#ident.erase();
        });
    }
    quote! {
        impl #impl_generics stylus_sdk::storage::Erase for #name #ty_generics #where_clause {
            fn erase(&mut self) {
                #erase_fields
            }
        }
    }
    .into()
}
