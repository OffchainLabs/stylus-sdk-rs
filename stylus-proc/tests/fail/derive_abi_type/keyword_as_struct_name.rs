// Copyright 2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

//! Compilation will fail if the type is not wrapped in the [`sol!`][alloy_sol_types::sol] macro.

use stylus_proc::AbiType;
use stylus_sdk::storage::StorageBool;

#[derive(AbiType)]
struct uint256 {
    bar: StorageBool,
}

fn main() {}
