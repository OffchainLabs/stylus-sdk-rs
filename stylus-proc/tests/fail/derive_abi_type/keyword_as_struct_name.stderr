error: struct name cannot be a Solidity keyword: `uint256`
  --> tests/fail/derive_abi_type/keyword_as_struct_name.rs:10:8
   |
10 | struct uint256 {
   |        ^^^^^^^

warning: type `uint256` should have an upper camel case name
  --> tests/fail/derive_abi_type/keyword_as_struct_name.rs:10:8
   |
10 | struct uint256 {
   |        ^^^^^^^ help: convert the identifier to upper camel case (notice the capitalization): `Uint256`
   |
   = note: `#[warn(non_camel_case_types)]` on by default
